File: electron/patcher/patchList.ts
import axios from 'axios';

export interface PatchInfo {
    index: number;
    filename: string;
}

/**
 * Fetches and parses a patch list file (plist.txt)
 * Format: Each line contains "<index> <filename>"
 */
export async function fetchPatchList(plistUrl: string): Promise<PatchInfo[]> {
    try {
        const response = await axios.get(plistUrl, {
            responseType: 'text',
            timeout: 30000
        });

        const lines = response.data.split('\n').filter((line: string) => line.trim());
        const patches: PatchInfo[] = [];

        for (const line of lines) {
            const trimmed = line.trim();
            if (!trimmed || trimmed.startsWith('//') || trimmed.startsWith('#')) {
                continue; // Skip empty lines and comments
            }

            // Format: "index filename" or just "filename"
            const parts = trimmed.split(/\s+/);
            if (parts.length >= 2) {
                const index = parseInt(parts[0], 10);
                const filename = parts[1];
                if (!isNaN(index) && filename) {
                    patches.push({ index, filename });
                }
            } else if (parts.length === 1) {
                // Just filename, use array index
                patches.push({ index: patches.length, filename: parts[0] });
            }
        }

        // Sort by index
        patches.sort((a, b) => a.index - b.index);

        return patches;
    } catch (error) {
        if (axios.isAxiosError(error)) {
            if (error.response?.status === 404) {
                // No patch list found, assume up to date
                return [];
            }
            throw new Error(`Failed to fetch patch list: ${error.message}`);
        }
        throw error;
    }
}

/**
 * Reads a local cache file to determine which patches have been applied
 */
export function getAppliedPatches(cachePath: string): number[] {
    try {
        const fs = require('fs');
        if (fs.existsSync(cachePath)) {
            const content = fs.readFileSync(cachePath, 'utf-8');
            const cache = JSON.parse(content);
            return cache.appliedPatches || [];
        }
    } catch {
        // Ignore errors
    }
    return [];
}

/**
 * Save the applied patches to cache
 */
export function saveAppliedPatches(cachePath: string, patches: number[]): void {
    const fs = require('fs');
    const cache = { appliedPatches: patches, lastUpdated: new Date().toISOString() };
    fs.writeFileSync(cachePath, JSON.stringify(cache, null, 2));
}

/**
 * Filter patches to get only unapplied ones
 */
export function getUnappliedPatches(allPatches: PatchInfo[], appliedIndices: number[]): PatchInfo[] {
    const appliedSet = new Set(appliedIndices);
    return allPatches.filter(patch => !appliedSet.has(patch.index));
}

File: src-tauri/src/patcher/patch_list.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PatchInfo {
    pub index: u32,
    pub filename: String,
}

pub async fn fetch_patch_list(plist_url: &str) -> Result<Vec<PatchInfo>, String> {
    let client = reqwest::Client::new();
    
    let response = client
        .get(plist_url)
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await
        .map_err(|e| format!("Failed to fetch patch list: {}", e))?;

    if response.status() == 404 {
        // No patch list found, assume up to date
        return Ok(vec![]);
    }

    if !response.status().is_success() {
        return Err(format!("Failed to fetch patch list: HTTP {}", response.status()));
    }

    let text = response
        .text()
        .await
        .map_err(|e| format!("Failed to read response: {}", e))?;

    let mut patches = Vec::new();

    for line in text.lines() {
        let trimmed = line.trim();
        
        // Skip empty lines and comments
        if trimmed.is_empty() || trimmed.starts_with("//") || trimmed.starts_with('#') {
            continue;
        }

        let parts: Vec<&str> = trimmed.split_whitespace().collect();

        if parts.len() >= 2 {
            // Format: "index filename"
            if let Ok(index) = parts[0].parse::<u32>() {
                patches.push(PatchInfo {
                    index,
                    filename: parts[1].to_string(),
                });
            }
        } else if parts.len() == 1 {
            // Just filename, use array index
            patches.push(PatchInfo {
                index: patches.len() as u32,
                filename: parts[0].to_string(),
            });
        }
    }

    // Sort by index
    patches.sort_by_key(|p| p.index);

    Ok(patches)
}

pub fn _get_applied_patches(cache_path: &str) -> Vec<u32> {
    match std::fs::read_to_string(cache_path) {
        Ok(content) => {
            if let Ok(cache) = serde_json::from_str::<serde_json::Value>(&content) {
                if let Some(applied) = cache.get("appliedPatches").and_then(|v| v.as_array()) {
                    return applied
                        .iter()
                        .filter_map(|v| v.as_u64().map(|n| n as u32))
                        .collect();
                }
            }
            vec![]
        }
        Err(_) => vec![],
    }
}

pub fn _save_applied_patches(cache_path: &str, patches: &[u32]) -> Result<(), String> {
    let cache = serde_json::json!({
        "appliedPatches": patches,
        "lastUpdated": chrono::Utc::now().to_rfc3339()
    });

    std::fs::write(cache_path, serde_json::to_string_pretty(&cache).unwrap())
        .map_err(|e| format!("Failed to save cache: {}", e))
}

pub fn _get_unapplied_patches(all_patches: &[PatchInfo], applied_indices: &[u32]) -> Vec<PatchInfo> {
    let applied_set: std::collections::HashSet<u32> = applied_indices.iter().copied().collect();
    
    all_patches
        .iter()
        .filter(|patch| !applied_set.contains(&patch.index))
        .cloned()
        .collect()
}
